native/src/main/java/io/ballerina/lib/data/yaml/io/BallerinaByteBlockInputStream.java

```diff
@@ -119,44 +86,32 @@ private boolean hasBytesInCurrentChunk() {
     }
 
     private boolean readNextChunk() throws InterruptedException {
-        Semaphore semaphore = new Semaphore(0);
-        Callback callback = new Callback() {
-
-            @Override
-            public void notifyFailure(BError bError) {
-                // Panic with an error
+        try {
+            Object result = env.getRuntime().call(iterator, nextMethodName);
+            if (result == null) {
                 done.set(true);
-                futureResultConsumer.accept(bError);
-                currentChunk = new byte[0];
-                semaphore.release();
-                // TODO : Should we panic here?
+                return true;
             }
-
-            @Override
-            public void notifySuccess(Object result) {
-                if (result == null) {
-                    done.set(true);
-                    currentChunk = new byte[0];
-                    semaphore.release();
-                    return;
-                }
-                if (result instanceof BMap<?, ?>) {
-                    BMap<BString, Object> valueRecord = (BMap<BString, Object>) result;
-                    final BString value = Arrays.stream(valueRecord.getKeys()).findFirst().get();
-                    final BArray arrayValue = valueRecord.getArrayValue(value);
-                    currentChunk = arrayValue.getByteArray();
-                    semaphore.release();
-                } else {
-                    // Case where Completes with an error
-                    done.set(true);
-                    semaphore.release();
-                }
+            if (result instanceof BMap<?, ?>) {
+                BMap<BString, Object> valueRecord = (BMap<BString, Object>) result;
+                final BString value = Arrays.stream(valueRecord.getKeys()).findFirst().get();
+                final BArray arrayValue = valueRecord.getArrayValue(value);
+                currentChunk = arrayValue.getByteArray();
+            } else {
+                // Case where Completes with an error
+                done.set(true);
```

Comment:
Can we directly return true here without using done temp variable

Reply:
Setting the value for `done` is required. I'll change the data type from `AtomicBoolean` to `boolean` as discussed externally.